src/test/deploy/functions/checkIam.spec.ts

```diff
@@ -253,4 +430,174 @@ describe("checkIam", () => {
       expect(setIamStub).to.have.been.calledWith(projectNumber, newIamPolicy, "bindings");
     });
   });
+
+  it("should add the pubsub publisher role for a new v2 storage function with v2 deployed functions", async () => {
+    const newIamPolicy = {
+      etag: "etag",
+      version: 3,
+      bindings: [
+        BINDING,
+        {
+          role: "roles/pubsub.publisher",
+          members: [`serviceAccount:${STORAGE_RES.email_address}`],
+        },
+      ],
+    };
+    storageStub.resolves(STORAGE_RES);
+    getIamStub.resolves({
+      etag: "etag",
+      version: 3,
+      bindings: [BINDING],
+    });
+    setIamStub.resolves(newIamPolicy);
+    const wantFn: backend.Endpoint = {
+      id: "wantFn",
+      entryPoint: "wantFn",
+      platform: "gcfv2",
+      eventTrigger: {
+        eventType: "google.cloud.storage.object.v1.finalized",
+        eventFilters: [
+          {
+            attribute: "bucket",
+            value: "my-bucket",
+          },
+        ],
+        retry: false,
+      },
+      ...SPEC,
+    };
+    const haveFn: backend.Endpoint = {
+      id: "haveFn",
+      entryPoint: "haveFn",
+      platform: "gcfv2",
+      eventTrigger: {
+        eventType: "google.firebase.firebasealerts.alerts.v1.published",
+        eventFilters: [
+          {
+            attribute: "alerttype",
+            value: "crashlytics.newFatalIssue",
+          },
+        ],
+        retry: false,
+      },
+      ...SPEC,
+    };
+
+    await checkIam.ensureServiceAgentRoles(projectNumber, backend.of(wantFn), backend.of(haveFn));
+
+    expect(storageStub).to.have.been.calledOnce;
+    expect(getIamStub).to.have.been.calledOnce;
+    expect(setIamStub).to.have.been.calledOnce;
+    expect(setIamStub).to.have.been.calledWith(projectNumber, newIamPolicy, "bindings");
+  });
+
+  it("should add the default bindings for a new v2 alerts function without v2 deployed functions", async () => {
```

Comment:
hm I seem to be missing something. Can you explain to me why PUBSUB TOKEN CREATOR binding isn't required when setting up firealerts functions?

Reply:
You mean the `roles/pubsub.publisher` role with the storage service account, right? That binding is only required for storage since the storage service account needs to be able to create events.